### PR TITLE
fix: fail on non-string MAP keys

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

fixes: https://github.com/confluentinc/ksql/issues/6181

Fixes ksqlDB so that statements that create maps with non-string keys fail.

Crux of the issue was in `SqlTypeParser`, which ignored the key type.

Additionally, there were many places in the code that checked for non-STRING keys, but they always were STRING keys by this point.  These checks have generally been removed in prep for supporting non-STRING keys.  It is the format that should decide if it can handle non-STRING keys or not, hence the code now fails these statements because the key/value format rejects them as unsupported. This opens the way to PROTOBUF supporting integer keys in the future: only the Protobuf Serde could should need to change. (https://github.com/confluentinc/ksql/issues/6177).

Main changes:

- Fix to `SqlMap` and factory method `SqlTypes.map` to have the map type take a key type as well as a value type.
- Fix `SqlTypeParser` so that the parser does not ignore the key type.
- Fix `MapType` of the `ParamType` family of types in the same was as `SqlMap`.
- Fix `UdfUtil` to handle map keys
- Fix exception types in a few places.

Most of the rest is just test code change to take into account the new parameters.  Sorry, forgot to push as separate commits.

### Testing done 

Usual

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-udf/src/test/java/io/confluent/ksql/schema/ksql/types/SqlArrayTest.java [hunk 1]
```diff
@@ -36,7 +36,7 @@ public void shouldImplementHashCodeAndEqualsProperly() {
     new EqualsTester()
         .addEqualityGroup(SqlArray.of(SOME_TYPE), SqlArray.of(SOME_TYPE))
         .addEqualityGroup(SqlArray.of(SqlPrimitiveType.of(SqlBaseType.BOOLEAN)))
-        .addEqualityGroup(SqlMap.of(SqlPrimitiveType.of(SqlBaseType.BOOLEAN)))
+        .addEqualityGroup(SqlMap.of(SqlTypes.STRING, SqlTypes.BOOLEAN))
```
Comment (vcrfxia) [inline ksqldb-udf/src/test/java/io/confluent/ksql/schema/ksql/types/SqlArrayTest.java:39]: This looks like it should be `SqlMap.of(SqlTypes.STRING, SqlPrimitiveType.of(SqlTypes.BOOLEAN))` instead?

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/UdtfLoaderTest.java [hunk 2]
```diff
@@ -97,13 +97,13 @@ public void shouldLoadParameterizedMapParams() {
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/UdtfLoaderTest.java:97]: Not your code but I notice that `shouldLoadParameterizedListParams()` (the test above this one in this file) doesn't test what it says it's testing. Wanna fix it while you're in here?

File: ksqldb-engine/src/test/java/io/confluent/ksql/integration/SecureIntegrationTest.java [hunk 3]
```diff
@@ -112,6 +114,10 @@
       .outerRule(Retry.of(3, ZooKeeperClientException.class, 3, TimeUnit.SECONDS))
       .around(TEST_HARNESS);
 
+
+  @Rule
+  public final Timeout timeout = Timeout.seconds(30);
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/integration/SecureIntegrationTest.java:119]: Curious: why the addition of a timeout to this test file in particular? Seems unrelated to the changes in this PR?

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/json/KsqlTypesSerdeModuleTest.java [hunk 4]
```diff
@@ -75,19 +78,22 @@ public void shouldSerDeSqlArrayTypes() throws JsonProcessingException {
...
         SqlTypes.DOUBLE,
         SqlTypes.STRING
-    };
+    );
 
-    for (final SqlType type : types) {
-      // When:
-      final SqlType out = MAPPER.readValue(MAPPER.writeValueAsString(SqlMap.of(type)), SqlType.class);
+    for (final SqlType keyType : types) {
+      for (final SqlType valueType : Lists.reverse(types)) {
```
Comment (vcrfxia) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/json/KsqlTypesSerdeModuleTest.java:89]: Any particular reason this is reversed?

**Review by vcrfxia — APPROVED** [review]
LGTM! Some questions inline.

File: ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestExecutorUtil.java [hunk 5]
```diff
@@ -308,15 +308,20 @@ private static ExecuteResultAndSources executePlan(
...
+      );
 
-    final Optional<List<DataSource>> dataSources = plan.getPlan().getQueryPlan()
-        .map(queryPlan -> getSources(queryPlan.getSources(), executionContext.getMetaStore()));
+      final Optional<List<DataSource>> dataSources = plan.getPlan().getQueryPlan()
+          .map(queryPlan -> getSources(queryPlan.getSources(), executionContext.getMetaStore()));
 
-    return new ExecuteResultAndSources(executeResult, dataSources);
+      return new ExecuteResultAndSources(executeResult, dataSources);
+    } catch (final KsqlStatementException e) {
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestExecutorUtil.java:321]: Nope, the code is attempting to catch a `KsqlStatementException` thrown by the engine, which will have the reformatted SQL in it, and throw a _almost_ matching `KsqlStatementException` that contains the pre-formatted sql.  Its a hack, but not a new one, this is copied for somewhere else in the test code.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/avro.json [hunk 6]
```diff
@@ -639,6 +639,27 @@
...
+      }
+    },
+    {
+      "name": "map with non-string keys fails - C*AS",
+      "statements": [
+        "CREATE STREAM INPUT (k INT, v DOUBLE) WITH (kafka_topic='input_topic', value_format='AVRO');",
+        "CREATE STREAM OUTPUT AS SELECT MAP(k:=v) FROM INPUT;"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlException",
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/avro.json:660]: See comment above for _why_

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/util/EntityUtilTest.java [hunk 7]
```diff
@@ -61,7 +61,7 @@ public void shouldBuildCorrectBooleanField() {
   public void shouldBuildCorrectMapField() {
     // Given:
     final LogicalSchema schema = LogicalSchema.builder()
-        .valueColumn(ColumnName.of("field"), SqlTypes.map(SqlTypes.INTEGER))
+        .valueColumn(ColumnName.of("field"), SqlTypes.map(SqlTypes.BIGINT, SqlTypes.INTEGER))
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/util/EntityUtilTest.java:64]: Correct, it's already called out in https://github.com/confluentinc/ksql/issues/6177

File: ksqldb-udf/src/test/java/io/confluent/ksql/schema/ksql/types/SqlArrayTest.java (hunk 1 above)
Comment (big-andy-coates) [reply ksqldb-udf/src/test/java/io/confluent/ksql/schema/ksql/types/SqlArrayTest.java:39]: No difference. Same thing.

File: ksqldb-engine/src/test/java/io/confluent/ksql/integration/SecureIntegrationTest.java (hunk 3 above)
Comment (big-andy-coates) [reply ksqldb-engine/src/test/java/io/confluent/ksql/integration/SecureIntegrationTest.java:119]: Is unrelated, but given this test deadlocked locally, I thought I'd just add the timeout.

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/json/KsqlTypesSerdeModuleTest.java (hunk 4 above)
Comment (big-andy-coates) [reply ksqldb-parser/src/test/java/io/confluent/ksql/parser/json/KsqlTypesSerdeModuleTest.java:89]: So they don't match.  Otherwise the code could be flipping key and value and the test would still pass.

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 8]
```diff
@@ -401,12 +401,14 @@ public Void visitCreateMapExpression(
...
-        final String types = keyTypes.stream()
-            .map(type -> type == null ? "NULL" : type.toString())
-            .collect(Collectors.joining(", ", "[", "]"));
+      if (keyTypes.stream().anyMatch(type -> type == null)) {
+        throw new KsqlException("Map keys can not be NULL");
+      }
 
-        throw new KsqlException("Only STRING keys are supported in maps but got: " + types);
+      if (keyTypes.size() == 0) {
+        throw new KsqlException("Cannot construct a map with all NULL key "
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java:409]: Removed.

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroDataTranslator.java [hunk 9]
```diff
@@ -135,7 +136,7 @@ private static Schema throwOnInvalidSchema(final Schema schema) {
       @Override
       public Void visitMap(final Schema schema, final Void key, final Void value) {
         if (schema.keySchema().type() != Type.STRING) {
-          throw new IllegalArgumentException("Avro only supports MAPs with STRING keys");
+          throw new KsqlException("Avro only supports MAPs with STRING keys");
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroDataTranslator.java:139]: On investigation, the `AvroFormat` wasn't validating the schema before this point. It does now and the validation code is now common.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/UdtfLoaderTest.java (hunk 2 above)
Comment (big-andy-coates) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/UdtfLoaderTest.java:97]: done

File: ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestExecutorUtil.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestExecutorUtil.java:321]: Resolved.

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroDataTranslator.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (vcrfxia) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroDataTranslator.java]: To check my understanding: you're saying this same code path is also used for validation, and that's why it's better to throw KsqlException than IllegalArgumentException?

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/json/KsqlTypesSerdeModuleTest.java (hunk 4 above)
Comment (vcrfxia) [reply ksqldb-parser/src/test/java/io/confluent/ksql/parser/json/KsqlTypesSerdeModuleTest.java:89]: I get why the additional for-loop was added (for that reason), but I'm wondering by there's `Lists.reverse()` in there. Is it because Guava contains a convenient method for copying a list and reversing its contents, but not for directly copying the list...?
